Linux/sgx/test_app/enclave/TestEnclave.cpp

```diff
@@ -90,184 +90,193 @@ struct evp_pkey_st {
+        return;
+    }
+    tbuf = buf;
+    i2d_PrivateKey(evp_pkey, &tbuf);
+
+    // print private key
+    printf ("{\"private\":\"");
+    for (i = 0; i < len; i++) {
+        printf("%02x", (unsigned char) buf[i]);
+    }
+    printf("\"}\n");
+
+    free(buf);
+
+    EVP_PKEY_free(evp_pkey);
 }
 
 void ec_key_gen()
 {
-	unsigned char entropy_buf[ADD_ENTROPY_SIZE] = {0};
-
-	RAND_add(entropy_buf, sizeof(entropy_buf), ADD_ENTROPY_SIZE);
-	RAND_seed(entropy_buf, sizeof(entropy_buf));
-
-	EC_KEY * ec = NULL;
-    int eccgroup;
-    eccgroup = OBJ_txt2nid("secp384r1");
-    ec = EC_KEY_new_by_curve_name(eccgroup);
-    if (ec == NULL) {
-    	printf("EC_KEY_new_by_curve_name failure: %ld\n", ERR_get_error());
-	    return;
+    EVP_PKEY_CTX * ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_EC, NULL);
+    if (!ctx)
+    {
+        printf("EVP_PKEY_CTX_new_id: %ld\n", ERR_get_error());
+        EVP_PKEY_CTX_free(ctx);
+        return;
     }
-    
-	EC_KEY_set_asn1_flag(ec, OPENSSL_EC_NAMED_CURVE);
-
-	int ret = EC_KEY_generate_key(ec);
-	if (!ret) {
-        printf("EC_KEY_generate_key failure\n");
-	    return;
-	}
-
-	EVP_PKEY *ec_pkey = EVP_PKEY_new();
-	if (ec_pkey == NULL) {
-		printf("EVP_PKEY_new failure: %ld\n", ERR_get_error());
-		return;
-	}
-	EVP_PKEY_assign_EC_KEY(ec_pkey, ec);
-
-	// public key - string
-	int len = i2d_PublicKey(ec_pkey, NULL);
-	unsigned char *buf = (unsigned char *) malloc (len + 1);
-	unsigned char *tbuf = buf;
-	i2d_PublicKey(ec_pkey, &tbuf);
-
-	// print public key
-	printf ("{\"public\":\"");
-	int i;
-	for (i = 0; i < len; i++) {
-	    printf("%02x", (unsigned char) buf[i]);
-	}
-	printf("\"}\n");
-
-	free(buf);
-
-	// private key - string
-	len = i2d_PrivateKey(ec_pkey, NULL);
-	buf = (unsigned char *) malloc (len + 1);
-	tbuf = buf;
-	i2d_PrivateKey(ec_pkey, &tbuf);
-
-	// print private key
-	printf ("{\"private\":\"");
-	for (i = 0; i < len; i++) {
-	    printf("%02x", (unsigned char) buf[i]);
-	}
-	printf("\"}\n");
-
-	free(buf);
-
-	EVP_PKEY_free(ec_pkey);
-	if (ec_pkey->pkey.ptr != NULL) {
-	  EC_KEY_free(ec);
-	}
+    int ret = EVP_PKEY_keygen_init(ctx);
+    if (!ret)
+    {
+        printf("EVP_PKEY_keygen_init: %ld\n", ERR_get_error());
+        EVP_PKEY_CTX_free(ctx);
+        return;
+    }
+    if (EVP_PKEY_CTX_set_ec_paramgen_curve_nid(ctx, NID_secp384r1) <= 0)
+    {
+        printf("EVP_PKEY_CTX_set_ec_paramgen_curve_nid: %ld\n", ERR_get_error());
+        EVP_PKEY_CTX_free(ctx);
+        return;
+    }
+    EVP_PKEY* ec_pkey = NULL;
+    if (EVP_PKEY_keygen(ctx, &ec_pkey) <= 0)
+    {
+        printf("EVP_PKEY_keygen: %ld\n", ERR_get_error());
+        EVP_PKEY_CTX_free(ctx);
+        return;
+    }
+    // public key - string
+    int len = i2d_PublicKey(ec_pkey, NULL);
+    unsigned char *buf = (unsigned char *) malloc (len + 1);
+    if (!buf)
+    {
+        printf("Failed in calling malloc()\n");
+        EVP_PKEY_CTX_free(ctx);
+        return;
+    }
+    unsigned char *tbuf = buf;
```

Comment:
This is a requirement from OpenSSL's i2d_PublicKey() and related APIs like i2d_x509(), according to:
https://www.openssl.org/docs/man1.1.1/man3/i2d_RSAPublicKey.html
if 

> to attempt to use a buffer directly:

> will result in buf apparently containing garbage because it was incremented after the call to point after the data just written. Also buf will no longer contain the pointer allocated by OPENSSL_malloc() and the subsequent call to OPENSSL_free() is likely to crash.

Reply:
Yes, you are right. Thanks.